### PR TITLE
specify cgroup ownership semantics

### PR DESCRIPTION
**Pull request by frasertweedale**
cgroups v2 supports secure delegation of cgroups.  Accordingly,
control over a cgroup (that is, creation of new child cgroups and
movement of processes and threads among the cgroup subtree exposed
to a container) can be safely delegated to a container.  Adjusting
the ownership enables real-world use cases like systemd-based
containers fully isolated in user namespaces.

To encourage adoption of this feature, and secure implementation,
define the semantics of cgroup ownership.  Changing/setting the
cgroup ownership is only allowed on cgroups v2, and the specific
files whose ownership can be change are mentioned.

In terms of current practice, this is already the behaviour of crun
(which also chown's the memory.oom.group file), and there is a pull
request for runc: https://github.com/opencontainers/runc/pull/3057
(the behaviour is enabled by an annotation).

### CODE REVIEW

File: config-linux.md
Comment (AkihiroSuda) [inline config-linux.md:209]: By default runtime MUST NOT change anything.
We need a new config field to opt-in to this.

Comment (frasertweedale) [reply config-linux.md:209]: That would put some existing implementations in immediately violation of the spec - `crun`, at least, already chowns the cgroup.  I'd appreciate more perspectives on this proposal.

Comment (AkihiroSuda) [reply config-linux.md:209]: But if we adopt the crun behavior, that would put runc in violation of the spec.

As runc is the reference implementation of the spec,  I suggest modifying this PR and crun implementation.

Comment (frasertweedale) [reply config-linux.md:209]: The proposal does not say runtimes MUST chown the cgroup - only that if they do, they MUST chown only the particular files mentioned.  I will update the language to clarify.

Comment (giuseppe) [reply config-linux.md]: by default `/sys/fs/cgroup` is mounted as read-only so in practice the cgroup ownership doesn't change much in what the container is allowed to do.  In order to use the `/sys/fs/cgroup` the mount must be writeable.

Maybe we can do the chowning when `/sys/fs/cgroup` is mounted as `rw`?  Otherwise it is unhelpful to have `/sys/fs/cgroup` writeable from an user not present in the current user namespace.

Comment (AkihiroSuda) [reply config-linux.md]: > Maybe we can do the chowning when /sys/fs/cgroup is mounted as rw?

SGTM :+1:
I think when the mount is RW && cgroup is v2, we can safely say that the runtime `SHOULD` chown the cgroup.

Comment (AkihiroSuda) [reply config-linux.md]: @kolyshkin @cyphar PTAL

Comment (frasertweedale) [reply config-linux.md]: @giuseppe @AkihiroSuda thanks for your suggestion.  I updated the text to specify that the runtime SHOULD chown the cgroup when the cgroupfs is mounted RW (and is cgroups v2), otherwise MUST NOT.

**Review by AkihiroSuda — APPROVED** [review]

File: config-linux.md
Comment (cyphar) [inline config-linux.md:207]: This should be a SHOULD NOT imho. While it might be a bad idea for security reasons, making this be something the spec explicitly disallows is a bit too extreme -- SHOULD NOT is strong enough, and including a justification is sufficient. There are reasons why you might want to do something different, and explicitly not allowing it a bit too extreme (not to mention adding MUST NOTs like this could be argued to be a SemVer-breaking change).

File: config-linux.md
Comment (cyphar) [inline config-linux.md:230]: Same comment about MUST and MUST NOT here. It should be SHOULD and SHOULD NOT. It's also entirely possible that it would be safe to allow the container to control the cgroup it's in (for instance if the runtime put the container in a nested cgroup and enforced the restrictions in the parent -- AFAIK this kind of thing is necessary to run systemd in containers so making it something you are not allowed to do is way too extreme IMHO).

File: config-linux.md
Comment (cyphar) [inline config-linux.md:218]: Same comment about MUST NOT here -- but in addition, the use of the phrase "if and only if" above makes this MUST NOT requirement a contradiction -- if I SHOULD do something if and only if X, then if not X I SHOULD NOT do it.

**Review by cyphar — CHANGES_REQUESTED** [review]
Seems reasonable but using MUSTs is far too strong IMHO.

**Comment by frasertweedale** [comment]
@cyphar thanks for your comments. I'll push an update later today, or tomorrow.

File: config-linux.md
Comment (kolyshkin) [inline config-linux.md:210]: nit: `s/to mounted/to be mounted/`

File: config-linux.md
Comment (frasertweedale) [reply config-linux.md]: relaxed to SHOULD.

File: config-linux.md
Comment (frasertweedale) [reply config-linux.md]: relaxed to SHOULD.

File: config-linux.md
Comment (frasertweedale) [reply config-linux.md]: relaxed to SHOULD.

File: config-linux.md
Comment (frasertweedale) [reply config-linux.md]: Fixed, thanks!

**Comment by frasertweedale** [comment]
PR updated to fix a grammar issue and relax MUST to SHOULD per @cyphar's comments.

**Comment by kolyshkin** [comment]
One question: should we also require cgroupns? It is assumed with cgroup v2, but what it it is not set in config?

I have just checked that by removing cgroup ns from config.json, and the whole host /sys/fs/cgroup is mounted (for both crun and runc). Meaning we definitely need to add cgroupns to the list of conditions.

**Comment by frasertweedale** [comment]
> One question: should we also require cgroupns? It is assumed with cgroup v2, but what it it is not set in config?
> 
> I have just checked that by removing cgroup ns from config.json, and the whole host /sys/fs/cgroup is mounted (for both crun and runc). Meaning we definitely need to add cgroupns to the list of conditions.

@kolyshkin good point, I agree.  Should we further restrict it to a cgroupns *without* `path` specified?  That is, where the runtime must create a new cgroupns?  Otherwise `path` could be a reference to an arbitrary, pre-existing cgroupns (possibly the "root" host cgroupns).

**Comment by frasertweedale** [comment]
Are there any further comments about this PR.  Any further concerns?  What else needs to be done for this PR to be ready to merge?

**Review by tianon — APPROVED** [review]
Seems sane to me

@cyphar @kolyshkin willing to take another look? 😅🙏

**Comment by frasertweedale** [comment]
Dear maintainers / reviewers, it would be great if we could get some more eyes on this and hopefully, approval and merge.
Ping @giuseppe @mrunalp ?

**Review by giuseppe — APPROVED** [review]
LGTM

**Comment by dqminh** [comment]
LGTM

**Comment by frasertweedale** [comment]
Thank you to reviewers.  We now have the requisite number of approvals.  @cyphar has an older review with changes requested; I believe I have resolved all of the feedback.

Can this be merged now?

**Comment by frasertweedale** [comment]
Weekly bump.  Could someone please merge it, or explain what prevents it?

**Comment by frasertweedale** [comment]
Weekly bump (Nov 22).   Could someone please merge it, or explain what else needs to be done or is preventing this PR from being merged?

**Comment by vbatts** [comment]
@frasertweedale thank you for the weekly bumps 😬 
Looks like the "MUST" that @cyphar commented on was never marked resolved. Perhaps that's why the other maintainers didn't merge? 🤷 

LGTM

**Comment by frasertweedale** [comment]
@vbatts thank you so much!  Thanks to all for their feedback and insights.

**Pull request merged**
